<a href="https://colab.research.google.com/github/Younes-Karimi/IST597-DataPrivacy/blob/main/DP_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<center><b>IST 597: Foundations of Data Privacy<br>
Spring 2021<br>
Homework #1<br>
Younes Karimi<br></center><br>
<left>Note: I have discussed with <b>Maya Merhi</b>, but we have done and submitted our works separately.

---



# Data Import

In [ ]:
import pandas as pd
import numpy as np
from google.colab import drive
import matplotlib.pyplot as plt
drive.mount('/content/drive')
dir = '/content/drive/My Drive/data_privacy/Adult/' # The original dataset directory
dir_numeric = '/content/drive/My Drive/data_privacy/Adult_numeric/' # The numerical version directory
pd.options.mode.chained_assignment = None  # default='warn'

# To run this code locally, you can comment out the directories above and provide absolute paths to your local directories as below:
# dir = 'PATH TO YOUR "Adult" folder' # The original dataset directory
# dir_numeric = 'PATH TO YOUR "Adult_numeric" folder' # The numerical version directory

In [51]:
# I read both the original and the converted numerical versions of the provided dataset. I show the basic information about the original data and continue my analyses with the numerical version for simplicity.
df = pd.read_table(dir+'dataset_full.txt', delim_whitespace=True, names=('age', 'education-num', 'work-class','marital-status', 'occupation', 'race', 'sex', 'native-country'))
dfn = pd.read_table(dir_numeric+'adults_numeric.del', delimiter=', ', engine='python', names=('age', 'work-class', 'education-num', 'marital-status', 'occupation', 'race', 'sex', 'native-country'))

# Original Dataset Analysis
List the column names in the dataset. For each column, identify the type of information (string, integer, etc.) and present the basic descriptives (range of values, mean, frequencies, outliers or unique values, etc). Explain the utility of the data that you want to preserve after your anonymization process. For example, discuss what types of queries you want your anonymized data to support. Your anonymization approach should support your answer.<br><br>
**???????????????????%%%%%%%%%%%%%%%%%%%% WE NEED TO EXPLAIN ABOUT UTILITY**

**???????????????????%%%%%%%%%%%%%%%%%%%% WE NEED TO EXPLAIN ABOUT UTILITY**

In [5]:
df.head()

age  education-num        work-class  ...   race   sex native-country
e1   47              3      Self-emp-inc  ...  White  Male           Cuba
e2   52             16  Self-emp-not-inc  ...  White  Male  United-States
e3   26              9           Private  ...  White  Male  United-States
e4   60              9           Private  ...  White  Male  United-States
e5   21             10           Private  ...  White  Male  United-States

[5 rows x 8 columns]

The table above shows the first entries in the dataset along with the header and column names.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30162 entries, e1 to e30162
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   education-num   30162 non-null  int64 
 2   work-class      30162 non-null  object
 3   marital-status  30162 non-null  object
 4   occupation      30162 non-null  object
 5   race            30162 non-null  object
 6   sex             30162 non-null  object
 7   native-country  30162 non-null  object
dtypes: int64(2), object(6)
memory usage: 2.1+ MB


This dataset consists of 8 columns and 30162 entries as shown above. The first two columns, *age* and *education_num*, are numerical data and have integer values while the rest of the columns are categorical data and have object types with string values.

In [7]:
print('Min age: {}\nMax age: {}'.format(df['age'].values.min(), df['age'].values.max()))
print('Mean age: %.2f' % (df['age'].values.sum() / df['age'].values.size))
print('\nSorted unique values: \n', np.sort(df['age'].unique()))
print('\nAge frequencies:\n', df['age'].value_counts(sort=True))

Min age: 17
Max age: 90
Mean age: 38.44

Sorted unique values: 
 [17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40
 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64
 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 88 90]

Age frequencies:
 36    852
31    851
33    837
34    836
35    828
     ... 
82      7
83      5
88      3
85      3
86      1
Name: age, Length: 72, dtype: int64


In [27]:
print('Min education number: {} year(s)\nMax education number: {} year(s)'.format(df['education-num'].values.min(), df['education-num'].values.max()))
print('Mean education number: %.2f year(s)' % (df['education-num'].values.sum() / df['education-num'].values.size))
print('\nEducation number frequencies:\n', df['education-num'].value_counts(sort=True))

Min education number: 1 year(s)
Max education number: 16 year(s)
Mean education number: 10.12 year(s)

Education number frequencies:
 9     9840
10    6678
13    5044
14    1627
11    1307
7     1048
12    1008
6      820
4      557
15     542
5      455
8      377
16     375
3      288
2      151
1       45
Name: education-num, dtype: int64


In [28]:
print('Work class frequencies:\n\n', df['work-class'].value_counts(sort=True))

Work class frequencies:

 Private             22286
Self-emp-not-inc     2499
Local-gov            2067
State-gov            1279
Self-emp-inc         1074
Federal-gov           943
Without-pay            14
Name: work-class, dtype: int64


In [29]:
print('Marital status frequencies:\n\n', df['marital-status'].value_counts(sort=True))

Marital status frequencies:

 Married-civ-spouse       14065
Never-married             9726
Divorced                  4214
Separated                  939
Widowed                    827
Married-spouse-absent      370
Married-AF-spouse           21
Name: marital-status, dtype: int64


In [30]:
print('Occupation frequencies:\n\n', df['occupation'].value_counts(sort=True))

Occupation frequencies:

 Prof-specialty       4038
Craft-repair         4030
Exec-managerial      3992
Adm-clerical         3721
Sales                3584
Other-service        3212
Machine-op-inspct    1966
Transport-moving     1572
Handlers-cleaners    1350
Farming-fishing       989
Tech-support          912
Protective-serv       644
Priv-house-serv       143
Armed-Forces            9
Name: occupation, dtype: int64


In [31]:
print('Race frequencies:\n\n', df['race'].value_counts(sort=True))

Race frequencies:

 White                 25933
Black                  2817
Asian-Pac-Islander      895
Amer-Indian-Eskimo      286
Other                   231
Name: race, dtype: int64


In [32]:
print('Sex frequencies:\n\n', df['sex'].value_counts(sort=True))

Sex frequencies:

 Male      20380
Female     9782
Name: sex, dtype: int64


In [33]:
print('Native country frequencies:\n\n', df['native-country'].value_counts(sort=True))

Native country frequencies:

 United-States                 27504
Mexico                          610
Philippines                     188
Germany                         128
Puerto-Rico                     109
Canada                          107
El-Salvador                     100
India                           100
Cuba                             92
England                          86
Jamaica                          80
South-Africa                     71
China                            68
Italy                            68
Dominican-Republic               67
Vietnam                          64
Guatemala                        63
Japan                            59
Poland                           56
Columbia                         56
Taiwan                           42
Iran                             42
Haiti                            42
Portugal                         34
Nicaragua                        33
Peru                             30
Greece                           2

In this section we presented the basic information about the original datatset provided which is a combination of numerical and categorical variables. But, because the numeric version of the dataset is available, for simplicity, we can perform our analysis with the numerical values and convert those numbers back to categorical values if needed at the end. In the next section, I perform a similar analysis with the numerical values.

# Numerical Dataset Analysis

In [7]:
dfn.head()

age  work-class  education-num  ...  race  sex  native-country
0   39           4             13  ...     0    0              0,
1   50           1             13  ...     0    0              0,
2   38           5              9  ...     0    0              0,
3   53           5              7  ...     1    0              0,
4   28           5             13  ...     1    1              2,

[5 rows x 8 columns]

As you can see in the table above, the read function and delimiter that we used for reading the dataset has not removed the tailing comma after the *native-contry* column. Therefore, the type of. last column is object instead of int64.

In [8]:
dfn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30162 entries, 0 to 30161
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   work-class      30162 non-null  int64 
 2   education-num   30162 non-null  int64 
 3   marital-status  30162 non-null  int64 
 4   occupation      30162 non-null  int64 
 5   race            30162 non-null  int64 
 6   sex             30162 non-null  int64 
 7   native-country  30162 non-null  object
dtypes: int64(7), object(1)
memory usage: 1.8+ MB


In [52]:
dfn['native-country']

0        0,
1        0,
2        0,
3        0,
4        2,
         ..
30157    0,
30158    0,
30159    0,
30160    0,
30161    0,
Name: native-country, Length: 30162, dtype: object

We replace the last column with a new column by removing the redundant comma and converting the type of column to int64. 

In [53]:
dfn['native-country'] = dfn['native-country'].replace({',':''}, regex=True).astype(int)
dfn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30162 entries, 0 to 30161
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   age             30162 non-null  int64
 1   work-class      30162 non-null  int64
 2   education-num   30162 non-null  int64
 3   marital-status  30162 non-null  int64
 4   occupation      30162 non-null  int64
 5   race            30162 non-null  int64
 6   sex             30162 non-null  int64
 7   native-country  30162 non-null  int64
dtypes: int64(8)
memory usage: 1.8 MB


In [15]:
print('Min age: {}\nMax age: {}'.format(dfn['age'].values.min(), dfn['age'].values.max()))
print('Mean age: %.2f' % (dfn['age'].values.sum() / dfn['age'].values.size))
print('\nSorted unique values: \n', np.sort(dfn['age'].unique()))
print('\nAge frequencies:\n', dfn['age'].value_counts(sort=True))

Min age: 17
Max age: 90
Mean age: 38.44

Sorted unique values: 
 [17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40
 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64
 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 88 90]

Age frequencies:
 36    852
31    851
33    837
34    836
35    828
     ... 
82      7
83      5
88      3
85      3
86      1
Name: age, Length: 72, dtype: int64


In [17]:
print('Work class frequencies:\n\n', dfn['work-class'].value_counts(sort=True))

Work class frequencies:

 5    22286
1     2499
3     2067
4     1279
0     1074
2      943
6       14
Name: work-class, dtype: int64


In [16]:
print('Min education number: {} year(s)\nMax education number: {} year(s)'.format(dfn['education-num'].values.min(), dfn['education-num'].values.max()))
print('Mean education number: %.2f year(s)' % (dfn['education-num'].values.sum() / dfn['education-num'].values.size))
print('\nEducation number frequencies:\n', dfn['education-num'].value_counts(sort=True))

Min education number: 1 year(s)
Max education number: 16 year(s)
Mean education number: 10.12 year(s)

Education number frequencies:
 9     9840
10    6678
13    5044
14    1627
11    1307
7     1048
12    1008
6      820
4      557
15     542
5      455
8      377
16     375
3      288
2      151
1       45
Name: education-num, dtype: int64


In [19]:
print('Marital status frequencies:\n\n', dfn['marital-status'].value_counts(sort=True))

Marital status frequencies:

 0    14065
6     9726
3     4214
5      939
2      827
4      370
1       21
Name: marital-status, dtype: int64


In [21]:
print('Occupation frequencies:\n\n', dfn['occupation'].value_counts(sort=True))

Occupation frequencies:

 1     4038
9     4030
0     3992
10    3721
3     3584
7     3212
11    1966
5     1572
6     1350
12     989
2      912
4      644
8      143
13       9
Name: occupation, dtype: int64


In [23]:
print('Race frequencies:\n\n', dfn['race'].value_counts(sort=True))

Race frequencies:

 0    25933
1     2817
2      895
3      286
4      231
Name: race, dtype: int64


In [24]:
print('Sex frequencies:\n\n', dfn['sex'].value_counts(sort=True))

Sex frequencies:

 0    20380
1     9782
Name: sex, dtype: int64


In [25]:
print('Native country frequencies:\n\n', dfn['native-country'].value_counts(sort=True))

Native country frequencies:

 0     27504
38      610
20      188
26      128
10      109
1       107
4       100
15      100
2        92
24       86
8        80
40       71
31       68
13       68
3        67
23       64
5        63
17       59
32       56
36       56
6        42
21       42
16       42
33       34
9        33
39       30
27       29
37       27
25       27
30       24
14       19
11       18
12       18
18       17
22       17
35       16
19       14
29       13
7        12
34       11
28        1
Name: native-country, dtype: int64


# Identification of Sensitive Information
Based on the data analysis you did, identify the explicit identifiers, quasi-identifiers, sensitive attributes and non-sensitive attributes in the dataset. List these attributes with your reasons for your classification.

**Identifiers**: There is not any explicit identifiers in this dataset as none of the attributes are unique and not any inndividual attribute can be used to uniquely identify each entry.
<br>
**Quasi-identifiers**: QIs can be identified by removing each sensitive attribute at a time and combining the rest of attributes in an ascending order (one attribute, combination of 2, 3, ...). For instance, for the person with age 86, because that is the only entry with this age, only the value of age would be enough to identify the rest of attribute for that person. But, for other attributes, as we combine more attributes and consider them collectively as quasi-identifiers, we might be able to uniquely identify more individuals having some combinations that no one else has them.
<br>
**Sensitive attributes**: While all these attributes may be considered sensitive and confidential depending on different scenarios, *age*, *occupation*, and *native country* might be more sensitive due to the low-frequent values and the outliers each of them have. For instance, there is only one person in the dataset with the age of 86 and that person might be uniquely identifiable if does not get suppressed or somehow anonymized.
<br>
**Non-sensitive attributes**: The rest of attributes might be less sensitive due to their more variety, but still can be confidential if they can be identified for specific individuals and may put them or their carriers into risk.

# Development of a K-Anonymization Approach
Develop a k-anonymization algorithm for the dataset. You can use any technique or combination of algorithms discussed in the class (k-anonymity, l-diversity, t-closeness, etc.). See below for a list of recommended algorithms. You MUST explain what your developed algorithm does at each step and how it anonymizes your specific dataset with respect to the explicit identifiers, quasi-identifiers, sensitive attributes, and non-sensitive data. Your code should be self-explanatory with understandable variable names, function names and comments. You MUST explicitly mention the chosen values of k in k-anonymity, l in l-diversity and t in t-closeness as per your algorithm of choice. Please consider if any outliers are present in the data. Your goal is to identify a high(er) value of k.

In [49]:
def find_value_ranges(df, group, scale=None):
  """
  This function identifies ranges of values in each column and in each specific
  group of entries
  """
  value_ranges = {}
  for col in df.columns:
    # Calculate the range of. values for each column in a specific group of 
    # entries by substracting max and min and adding one to include the first 
    # (min) value as well
    value_range = df[col][group].max() - df[col][group].min() + 1
    if scale is not None:
      value_range = value_range / scale[col]
    value_ranges[col] = value_range
  return value_ranges

The parameter k for *k-anonymity* can be set here. Here we have used 4-anonymity and *k=4*.

In [50]:
K = 4

In [54]:
ranges = find_value_ranges(dfn, list(dfn.index))
ranges

{'age': 74,
 'education-num': 16,
 'marital-status': 7,
 'native-country': 41,
 'occupation': 14,
 'race': 5,
 'sex': 2,
 'work-class': 7}

In [14]:
def split(df, group, col):
  dfg = df[col][group] # Part of df that intersects the group and the column
  dfl = dfg.index[dfg < dfg.median()] # Lower values of the column in that the group
  dfr = dfg.index[dfg >= dfg.median()] # Higher values of the column in that the group
  return (dfl, dfr)

In [18]:
def group_entries(df, quasi_identifiers, scale):
  """
  This function groups the entries together based on the QIs
  """
  final_groups = []
  initial_groups = [df.index]
  while initial_groups:
    group = initial_groups.pop(0)
    value_ranges = find_value_ranges(df[quasi_identifiers], group, scale)
    for col, span in sorted(value_ranges.items(), key=lambda x:-x[1]):
      lp, rp = split(df, group, col)
      if len(lp)<K or len(rp)<K:
        continue
      initial_groups.extend((lp, rp))
      break
    else:
        final_groups.append(group)
  return final_groups

In [57]:
sensitive_col = 'occupation'
quasi_identifiers = ['age', 'education-num', 'race', 'sex', 'work-class', 'marital-status', 'native-country']
groups = group_entries(dfn, quasi_identifiers, ranges)

In [58]:
len(groups)

1521

In [42]:
def agg_column(series):
  return [series.mean()]

In [43]:
def anonymize_dataset(df, groups, quasi_identifiers, sensitive_col, max_groups=None):
  aggregations = {}
  for column in quasi_identifiers:
    aggregations[column] = agg_column
  rows = []
  for i, group in enumerate(groups):
    if max_groups is not None and i > max_groups:
      break
    grouped_columns = df.loc[group].agg(aggregations, squeeze=False)
    sensitive_counts = df.loc[group].groupby(sensitive_col).agg({sensitive_col: 'count'})
    values = grouped_columns.iloc[0].to_dict()
    for sensitive_value, count in sensitive_counts[sensitive_col].items():
      if count == 0:
        continue
      values.update({sensitive_col: sensitive_value, 'count': count})
      rows.append(values.copy())
  return pd.DataFrame(rows)

In [59]:
dfn = anonymize_dataset(dfn, groups, quasi_identifiers, sensitive_col)

In [62]:
dfn.sort_values(quasi_identifiers+[sensitive_col]).head(20)

age  education-num      race  ...  native-country  occupation  count
1640  17.0       5.666667  0.166667  ...        0.000000           4      1
1641  17.0       5.666667  0.166667  ...        0.000000           7      4
1642  17.0       5.666667  0.166667  ...        0.000000          12      1
2954  17.0       7.000000  0.250000  ...        0.000000           1      1
2955  17.0       7.000000  0.250000  ...        0.000000           3      1
2956  17.0       7.000000  0.250000  ...        0.000000           4      1
2957  17.0       7.000000  0.250000  ...        0.000000           9      1
123   17.0       7.162791  0.116279  ...        0.546512           1      1
124   17.0       7.162791  0.116279  ...        0.546512           3     36
125   17.0       7.162791  0.116279  ...        0.546512           4      1
126   17.0       7.162791  0.116279  ...        0.546512           5      1
127   17.0       7.162791  0.116279  ...        0.546512           6      2
128   17.0       7.162791  0.116279  ...        0.546512           7     26
129   17.0       7.162791  0.116279  ...        0.546512           8      6
130   17.0       7.162791  0.116279  ...        0.546512           9      1
131   17.0       7.162791  0.116279  ...        0.546512          10     12
2958  17.0       7.166667  0.000000  ...        1.666667           1      2
2959  17.0       7.166667  0.000000  ...        1.666667           7      1
2960  17.0       7.166667  0.000000  ...        1.666667          10      3
114   17.0       7.188235  0.141176  ...        0.823529           1      3

[20 rows x 9 columns]

# Analysis
Explain how your solution impacts the utility of the dataset. Does it retain the original data distribution. Provide at least two queries as test examples to show the differences in utility before and after the anonymization.

As it can be seen in the dataset analysis section, there is only one person with *age* = 86 and one person with *native-country* = *Holand-Netherlands*. Here I have listed the corresponding entries in the original dataset and similar queries to retrieve. them from the anonymized dataset. But, because they were outliers, they have. been suppressed in the anonymized dataset and no. entries with these values can be found in *dfn*. This suppression not only removes some entries from the dataset but also may change the. range of some attributes like *age* and *native-country* in these cases.

In [81]:
df.loc[df['age'] == 86]

age  education-num work-class  ...   race     sex native-country
e27739   86             14    Private  ...  White  Female  United-States

[1 rows x 8 columns]

In [78]:
dfn.loc[dfn['age'] == 86]

Empty DataFrame
Columns: [age, education-num, race, sex, work-class, marital-status, native-country, occupation, count]
Index: []

In [82]:
df.loc[df['native-country'] == 'Holand-Netherlands']

age  education-num work-class  ...   race     sex      native-country
e21177   32             10    Private  ...  White  Female  Holand-Netherlands

[1 rows x 8 columns]

In [79]:
dfn.loc[dfn['native-country'] == 'Holand-Netherlands']

Empty DataFrame
Columns: [age, education-num, race, sex, work-class, marital-status, native-country, occupation, count]
Index: []

# References
[1] [An implementation of k-anonymity algorithm](https://github.com/Nuclearstar/K-Anonymity/blob/master/k-Anonymity.ipynb)<br>
[2] LeFevre, K., DeWitt, D. J., & Ramakrishnan, R. (2006, April). [Mondrian multidimensional k-anonymity](https://ieeexplore.ieee.org/abstract/document/1617393?casa_token=0zDsl8OE5sAAAAAA:YSvNLq5wADBKGkcteunAvwqAS8E2NS6ekvURtv7P2-WXMaEiTUOsMWw9KgtEBFIkxIf8gZ7blw). In 22nd International conference on data engineering (ICDE'06) (pp. 25-25). IEEE.